Source/x11/XGServerWindow.m

```diff
@@ -4409,75 +4415,140 @@ - (void) freecursor: (void*) cid
+
+      for (i = 0; i < monitorsCount; i++)
+        {
+          output_info = XRRGetOutputInfo(dpy, screen_res, screen_res->outputs[i]);
+          monitors[i].screen_id = defScreen;
+          if (output_info->crtc)
+            {
+              crtc_info = XRRGetCrtcInfo(dpy, screen_res, output_info->crtc);
+
+              // Fill the cache of device parameters
+              monitors[i].depth = [self windowDepthForScreen: defScreen];
+              monitors[i].resolution = [self resolutionForScreen: defScreen];
+              monitors[i].frame = NSMakeRect(crtc_info->x, crtc_info->y,
+                                             crtc_info->width, crtc_info->height);
+              // Add number of device
+              [tmpScreens addObject: [NSNumber numberWithInt: i]];
+            }
+        }
+      XRRFreeScreenResources(screen_res);
+      screens = [NSArray arrayWithArray: tmpScreens];
+    }
+#endif
+
+  if (monitorsCount == 0)
+    {
+      /* Assumed that it's always available 1 screen per application. We only
+         need to know it's number and it was saved in _initXContext as
+         `defScreen`. */
+      monitorsCount = 1;
+      monitors = NSZoneMalloc([self zone], sizeof(MonitorDevice));
+      monitors[0].screen_id = defScreen;
+      monitors[0].depth = [self windowDepthForScreen: defScreen];
+      monitors[0].resolution = [self resolutionForScreen: defScreen];
+      monitors[0].frame = NSMakeRect(0, 0,
+                                     DisplayWidth(dpy, defScreen),
+                                     DisplayHeight(dpy, defScreen));
+      screens = [NSArray arrayWithObject: [NSNumber numberWithInt: defScreen]];
+    }
+  
+  return screens;
 }
 
-- (NSWindowDepth) windowDepthForScreen: (int) screen_num
+// `screen` is a Xlib screen number.
+- (NSWindowDepth) windowDepthForScreen: (int)screen
 {
-  Screen	*screen;
+  Screen	*x_screen;
   int		 class = 0, bpp = 0;
 
-  screen = XScreenOfDisplay(dpy, screen_num);
-  if (screen == NULL)
+  if (screen < 0 || screen >= monitorsCount)
+    {
+      return 0;
+    }
+  
+  x_screen = XScreenOfDisplay(dpy, screen);
+  
+  if (x_screen == NULL)
     {
       return 0;
     }
 
-  bpp = screen->root_depth;
-  class = screen->root_visual->class;
+  bpp = x_screen->root_depth;
+  class = x_screen->root_visual->class;
 
   return _computeDepth(class, bpp);
 }
 
-- (const NSWindowDepth *) availableDepthsForScreen: (int) screen_num
+// `screen` is a monitor index not X11 screen
+- (const NSWindowDepth *) availableDepthsForScreen: (int)screen
 {
-  Screen	*screen;
+  Screen	*x_screen;
   int		 class = 0;
   int		 index = 0;
   int		 ndepths = 0;
   NSZone	*defaultZone = NSDefaultMallocZone();
   NSWindowDepth	*depths = 0;
 
-  if (dpy == NULL)
+  if (dpy == NULL || screen < 0 || screen >= monitorsCount)
     {
       return NULL;
     }
 
-  screen = XScreenOfDisplay(dpy, screen_num);
-  if (screen == NULL)
+  x_screen = XScreenOfDisplay(dpy, monitors[screen].screen_id);
```

Comment:
How could that in your code ever be different from the default screen?

Reply:
You're right. Let's leave this as is for now. I'll remove it later if we decide to remove gswindow_device_t->screen_id.